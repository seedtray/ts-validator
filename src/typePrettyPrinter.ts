/**
 * Pretty printer for parsed types.
 */
import {
    ArrayType,
    EnumType,
    IntersectionType,
    LiteralBooleanType,
    LiteralNumberType,
    LiteralStringType, NamedType,
    ObjectType,
    PrimitiveType,
    TupleType,
    Type,
    TypeVisitor,
    UnionType,
} from './types'

import {fail} from './errors'

/**
 * A prettified type fragment that fits in a single line.
 */
export class PrettyInline {

    static joinSurround(parts: string[],
                        joiner: string,
                        prefix: string = '',
                        suffix: string = ''): PrettyInline {

        return new PrettyInline(prefix + parts.join(joiner) + suffix)
    }

    constructor(public expression: string) {
    }

    toLines(indent: number): string {
        return ' '.repeat(indent) + this.expression
    }

    toString(): string {
        return this.toLines(0)
    }

    withSuffix(suffix: string): PrettyInline {
        return new PrettyInline(this.expression + suffix)
    }
}

/**
 * A prettified block of text for representing a type.
 */
export class PrettyBlock {
    constructor(public opening: PrettyInline, public parts: PrettyType[], public end: PrettyInline) {

    }

    surround(prefix: string, suffix: string): PrettyBlock {
        return new PrettyBlock(
            new PrettyInline(`${prefix}${this.opening.expression}`),
            this.parts,
            new PrettyInline(`${this.end.expression}${suffix}`)
        )
    }

    toLines(indent: number): string {
        return [
            this.opening.toLines(indent),
            this.parts.map(part => part.toLines(indent + 2)).join('\n'),
            this.end.toLines(indent),
        ].join('\n')
    }

    toString(): string {
        return this.toLines(0)
    }
}

type PrettyType = PrettyInline | PrettyBlock

/**
 * Transform a parsed type to a pretty printed version of it.
 */
export class TypePrettyPrinter implements TypeVisitor<PrettyType> {

    visitPrimitive(p: PrimitiveType): PrettyType {
        return new PrettyInline(p.target)
    }

    visitArray(a: ArrayType): PrettyType {
        const prettyOf = a.target.accept(this)
        if (a.target instanceof PrimitiveType) {
            return new PrettyInline(`${prettyOf}[]`)
        } else if (prettyOf instanceof PrettyInline) {
            return new PrettyInline(`Array<${prettyOf}>`)
        } else if (prettyOf instanceof PrettyBlock) {
            return prettyOf.surround('Array<', '>')
        } else {
            return fail()
        }
    }

    visitObject(o: ObjectType): PrettyType {
        return new PrettyBlock(
            new PrettyInline('{'),
            Array.from(this.prettyProperties(o.properties)),
            new PrettyInline('}')
        )
    }

    visitUnion(u: UnionType): PrettyType {
        return this.prettyListOfTypes(
            u.target,
            parts =>
                new PrettyInline(parts.map(part => part.expression).join(' | ')),
            parts => new PrettyBlock(
                new PrettyInline('Union('),
                parts,
                new PrettyInline(')')
            )
        )
    }

    visitIntersection(i: IntersectionType): PrettyType {
        return this.prettyListOfTypes(
            i.target,
            parts =>
                new PrettyInline(parts.map(part => part.expression).join(' & ')),
            parts => new PrettyBlock(
                new PrettyInline('Intersection('),
                parts,
                new PrettyInline(')')
            )
        )
    }

    visitTuple(t: TupleType): PrettyType {
        return this.prettyListOfTypes(
            t.target,
            parts => PrettyInline.joinSurround(
                parts.map(part => part.expression),
                ', ', '[', ']'
            ),
            parts => new PrettyBlock(
                new PrettyInline('Union('),
                parts,
                new PrettyInline(')')
            )
        )

    }

    visitEnum(t: EnumType): PrettyType {
        return new PrettyBlock(
            new PrettyInline('Enum('),
            Array.from(this.prettyEnumMembers(t.members)),
            new PrettyInline(')'),
        )
    }

    visitLiteralString(literal: LiteralStringType): PrettyType {
        return new PrettyInline(`'${literal.value}'`)
    }

    visitLiteralNumber(literal: LiteralNumberType): PrettyType {
        return new PrettyInline(`'${literal.value}'`)
    }

    visitLiteralBoolean(literal: LiteralBooleanType): PrettyType {
        return new PrettyInline(`'${literal.value}'`)
    }

    visitNamedType(t: NamedType): PrettyType {
        const nameRef = ` [${t.name}]`
        const prettyType = t.target.accept(this)
        if (prettyType instanceof PrettyBlock) {
            return new PrettyBlock(
                prettyType.opening.withSuffix(nameRef),
                prettyType.parts,
                prettyType.end
            )
        } else {
            return prettyType.withSuffix(nameRef)
        }
    }

    private* prettyProperties(properties: Map<string, Type>): IterableIterator<PrettyType> {
        for (const [name, type] of properties.entries()) {
            const prettyProperty = type.accept(this)
            if (prettyProperty instanceof PrettyInline) {
                yield new PrettyInline(`${name}: ${prettyProperty.expression}`)
            } else if (prettyProperty instanceof PrettyBlock) {
                yield prettyProperty.surround(`${name}: `, '')
            }
        }
    }

    private prettyListOfTypes(listOfTypes: Type[], inlineCase: (parts: PrettyInline[]) => PrettyType,
                              complexCase: (parts: PrettyType[]) => PrettyType): PrettyType {
        const prettyMembers = listOfTypes.map(target => target.accept(this))
        if (prettyMembers.every(pretty => pretty instanceof PrettyInline)) {
            return inlineCase(prettyMembers as PrettyInline[])
        } else {
            return complexCase(prettyMembers)
        }

    }

    // Quote a string. Does not deal well with single quotes in values.
    private quote(s: string): string {
        return `'${s}'`
    }

    private* prettyEnumMembers(
        members: Map<string, string | number>): IterableIterator<PrettyInline> {

        for (const [name, value] of members.entries()) {
            const prettyValue = typeof value === 'string' ? this.quote(value) : value
            yield  new PrettyInline(`${name}=${prettyValue},`)

        }

    }

}
