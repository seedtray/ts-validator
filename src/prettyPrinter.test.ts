import 'jest'
import {TypePrettyPrinter} from 'typePrettyPrinter.ts'
import {
    ArrayType, booleanType, EnumType, IntersectionType, nullType, numberType, ObjectType,
    stringType, TupleType,
    undefinedType, UnionType
} from './types'

const pp = new TypePrettyPrinter()
test('pretty prints basic types', () => {
    expect(numberType.accept(pp).toString()).toBe('number')
    expect(booleanType.accept(pp).toString()).toBe('boolean')
    expect(stringType.accept(pp).toString()).toBe('string')
    expect(nullType.accept(pp).toString()).toBe('null')
    expect(undefinedType.accept(pp).toString()).toBe('undefined')
})
test('pretty prints tuples', () => {
    expect(TupleType.of([numberType, stringType]).accept(pp).toString())
        .toBe('[number, string]')
})
test('short pretty prints arrays', () => {
    expect(ArrayType.of(numberType).accept(pp).toString()).toBe('number[]')
})
test('prints more complex arrays', () => {
    expect(
        ArrayType.of(
            TupleType.of([numberType])
        ).accept(pp).toString()).toBe('Array<[number]>')
})
test('pretty print union types', () => {
    expect(UnionType.of([numberType, booleanType]).accept(pp).toString())
        .toBe('number | boolean')
})
test('pretty print intersection types', () => {
    expect(IntersectionType.of([numberType, booleanType]).accept(pp).toString())
        .toBe('number & boolean')
})
test('pretty prints enums', () => {
    const enumType = new EnumType()
    enumType.add('test', 1)
    enumType.add('another', 2)
    enumType.add('aString', 'test')
    expect(enumType.accept(pp).toString()).toBe(
        `Enum(\n  test=1,\n  another=2,\n  aString='test',\n)`)
})
test('pretty prints simple objects', () => {
    const ot = new ObjectType()
    ot.add('a', numberType)
    ot.add('b', TupleType.of([numberType, stringType]))
    expect(ot.accept(pp).toString()).toBe(
        `{\n  a: number\n  b: [number, string]\n}`
    )
})

test('pretty nested objects', () => {
    const nested = ObjectType.of({o2: nullType})
    const ot = ObjectType.of({
        a: numberType,
        b: TupleType.of([numberType, stringType]),
        ot2: nested,
        at: ArrayType.of(nested)
    })
    expect(ot.accept(pp).toString()).toBe([
        '{',
        '  a: number',
        '  b: [number, string]',
        '  ot2: {',
        '    o2: null',
        '  }',
        '  at: Array<{',
        '    o2: null',
        '  }>',
        '}',
    ].join('\n'))
})
