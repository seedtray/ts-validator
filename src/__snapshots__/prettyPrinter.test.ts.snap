// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pretty nested objects 1`] = `
"{
  a: number
  b: [number, string]
  ot2: {
    o2: null
  }
  at: Array<{
    o2: null
  }>
}"
`;

exports[`pretty prints enums 1`] = `
"Enum(
  test=1,
  another=2,
  aString='test',
)"
`;

exports[`pretty prints simple objects 1`] = `
"{
  a: number
  b: [number, string]
}"
`;

exports[`pretty recursive types 1`] = `
"{ [CompositionWithRecursive]
  t: { [Recursive33]
    d: { [Recursive31]
      a: { [Recursive32]
        b: RecursiveReference<Recursive32>
        c: RecursiveReference<Recursive33>
      }
    }
  }
}"
`;
